bftengine/src/bftengine/messages/MessageBase.cpp

```diff
@@ -46,11 +46,61 @@ void MessageBase::printLiveMessages() {
 namespace bftEngine {
 namespace impl {
 
+// static class members for diagnostics server
+std::array<std::atomic<size_t>, MsgCode::MaxMsgCodeVal> MessageBase::AliveIncomingExtrnMsgsBufs{};
```

Comment:
@hshomroni 
I beleive that this whole implementation can be much simplified and lock free:
1). define an std::map<MsgCode::MaxMsgCodeVal, atomic<size_t>>
2. initialize the map when replica goes up, currently we have approx N=25 types of messages. push 25 paies with value 0. after that you have a "static strucutre map" which allows updating only the values. you should never perform anywhere in the code after init part which is done in single thread mode any push/delete or modifications to the map keys.
3.to check if an arriveing type is valid and is im the map (must be in regular case, since you inserted all possible types) - accress map::at[msgType] ( std::map::at is considered thread safe according to c++ standard).
4. updating the value is atomic.
5. you've got a thrad safe map since it's structure never change. 
You dont need 2 data structures, just one, and you don't need a lock.
I beleive the suggested implementation is far simpler than current one, and thread safe (current one requeres lock as i've stated in another comment).
for clarification, let speak offline.

comment: I suggest using std::unordered_map with std::unordered_map::at. at is a const function and cannot modify the map.

Reply:
I started with the approach you suggest, including disabling the option to delete/erase keys. However, I did not find any way to iterate over an enum taking only the named values (for the init in step 2). That's why I decided to move to the array approach and avoid runtime keys insertions. I will add the static_assert you suggested to prevent the risk of exploding this array's size.